### PR TITLE
feat(editorconfig): reenable editorconfig support, add cli flag to disable it

### PR DESCRIPTION
**Pull request by dyc3**
## Summary

This PR reenables editorconfig support, and additionally adds a cli flag, `--use-editorconfig`, to control whether or not it's enabled.

related to: #1724
- [x] blocked by: #3260 

## Test Plan

I've reenabled the unit tests that I wrote for this.

### CODE REVIEW

**Review by ematipico — COMMENTED** [review]
In our repository, we have command called `pnpm check` that runs the Biome CLI in dev mode. If you run, you'll notice that it will break: 

```
configuration ━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ✖ Couldn't parse the **.yml, reason: recursive wildcards must form a single path component
  

 ELIFECYCLE  Command failed with exit code 1.
```

We should fix that

**Comment by ematipico** [comment]
After fixing the issue with editor config `**.yml` -> `*.yml` (we should make the error message a bit simpler to understand...), I run into overriding issues.. meaning that tnow `tsconfig.json` is failing because it isn't parsed as JSOC anymore:

```
~/packages/@biomejs/js-api/tsconfig.json:3:24 parse ━━━━━━━━━━━━━━━━━━━━━━━

  ✖ JSON standard does not allow comments.
  
    1 │ {
    2 │   "compilerOptions": {
  > 3 │     "target": "es2020" /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */,
      │                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    4 │     "module": "commonjs" /* Specify what module code is generated. */,
    5 │     "declaration": true /* Generate .d.ts files from TypeScript and JavaScript files in your project. */,
  
```

**Comment by dyc3** [comment]
I've updated it so that error will now be printed as:
```
/home/<absolute path>/.editorconfig configuration ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  ✖ Couldn't parse the pattern "**.yml". Reason: Pattern syntax error near position 2: recursive wildcards must form a single path component
```
and it'll now convert that pattern into one biome can use. That `tsconfig.json` thing sounds like it's out of scope for editorconfig support?

**Comment by ematipico** [comment]
>  That tsconfig.json thing sounds like it's out of scope for editorconfig support?

But it's a regression I am experiencing **in this PR**. The command `pnpm check` doesn't work anymore. Is it expected?

**Comment by dyc3** [comment]
I think this is a bug on `main`. Adding the following override to this repo's `biome.json` produces the same result.

```json
"overrides": [
    {
      "include": ["**/*.json"],
      "formatter": {
        "indentStyle": "space"
      }
    }
  ]
```

Obviously, we can't merge this without that being resolved. I'll look into it.

**Comment by dyc3** [comment]
Running `cargo run --bin biome -- check packages/@biomejs/backend-jsonrpc/tsconfig.json` with the override does not produce the diagnostics. I think the source of the problem has something to do with the caching being done here: 
https://github.com/biomejs/biome/blob/cc65fe857929896b88a40cfee759a1f80d3ab876/crates/biome_service/src/settings.rs#L1095-L1100

**Comment by dyc3** [comment]
@ematipico I've submitted a draft PR to fix this in #3260, but I don't really like the solution. Details are in the PR.

**Comment by dyc3** [comment]
@ematipico This PR should be unblocked now.

I went ahead and disabled the caching for json overrides because the behavior was incorrect, as discussed in #3260, despite the potential for performance regressions.

File: crates/biome_service/src/settings.rs
Comment (ematipico) [inline crates/biome_service/src/settings.rs:956]: I wonder if we could keep it, add a `_` at the beginning of its name, and add `#[allow(dead_code)]`. I am sure we will need it again once we fix the caching

File: crates/biome_service/src/configuration.rs
Comment (ematipico) [inline crates/biome_service/src/configuration.rs:283]: You can use `config.0.as_ref().and_then(|c| c.overrides)` to reduce the `Some` to one, and avoid too many indentations.

File: crates/biome_cli/src/cli_options.rs
Comment (ematipico) [inline crates/biome_cli/src/cli_options.rs:90]: I think users would also want to control the `.editoconfig` from the configuration file `biome.json`. If you think about it, this isn't related to the CLI (like colours, the verbosity of diagnostics, etc.) but it's something related to the formatter.

Maybe this option should be placed inside `formatter.useEditorconfig`, next to `formatter.enabled`.

File: crates/biome_configuration/src/formatter.rs
Comment (ematipico) [inline crates/biome_configuration/src/formatter.rs:106]: I am a bit worried about this default. It makes sense to have to `true`, however this could break some formatting for users that already use Biome and have an `.editorconfig` somewhere. What do you think? Should we set it `false` for now, and set it to `true` for `v2`?

File: crates/biome_configuration/src/formatter.rs
Comment (ematipico) [inline crates/biome_configuration/src/formatter.rs:21]: nit: The documentation should mention the default.

File: crates/biome_configuration/src/editorconfig.rs
Comment (ematipico) [inline crates/biome_configuration/src/editorconfig.rs:311]: Great workaround!

**Review by ematipico — APPROVED** [review]
This looks good to me. I am a bit worried about the default, and I would like to know what you think. Once settled on the default value, we can merge it :) Great work @dyc3, thank you for landing this feature

File: crates/biome_configuration/src/formatter.rs
Comment (dyc3) [reply crates/biome_configuration/src/formatter.rs:106]: That would be the most conservative. On the other hand, prettier enables editorconfig support by default, and keeping this behavior when migrating from prettier would be a better user experience.

I think we should set this to false by default, but when migrating from prettier, set it to true if it's unspecified.

ref: https://prettier.io/docs/en/configuration.html#editorconfig

Comment (dyc3) [reply crates/biome_configuration/src/formatter.rs:106]: actually it turns out that prettier's `options.editorconfig` doesn't work at all, see: https://github.com/prettier/prettier/issues/15255

Comment (ematipico) [reply crates/biome_configuration/src/formatter.rs]: Yeah, sounds good me!

**Review by ematipico — APPROVED** [review]

**Pull request merged**
